### ISSUE TITLE
Fvar table issue, related to the Variable Font Origin?

### ISSUE THREAD

**Issue by emmamarichal** [googlefonts/gftools#1069]
Hello!

I want this FVAR table: 
_(Playfair, Version 2.001;gftools[0.9.30])_

![Image](https://github.com/user-attachments/assets/e787c801-f0d6-4d67-a449-c142d81b539b)

But for the new version of Playfair, with the same parameters as in the previous version, I get:

![Image](https://github.com/user-attachments/assets/14c3d875-fa0a-4903-8f28-4db19b570bf7)

Here is the instances set up and the axis mapping didn't change:

![Image](https://github.com/user-attachments/assets/549d498b-f391-4429-ac52-0020ef4ec98f)

When I remove the `Variable Font Origin` parameter, the fvar table changed also. Is there anything that changed in gftools builder, that could explain this? Otherwise, do you have any idea so I can get back the previous instances?

Here you can find the two versions of files:
[2023 version.zip](https://github.com/user-attachments/files/17949776/2023.version.zip)
[2024 version.zip](https://github.com/user-attachments/files/17949788/2024.version.zip)

**Comment by simoncozens** [comment]
So I don't know what is happening, and I am debugging this as I go along. If you just want the answer, skip to the bottom.

For 2024, I get a good fvar table with fontmake:

```
    <!-- Light -->
    <NamedInstance flags="0x0" subfamilyNameID="263">
      <coord axis="opsz" value="12.0"/>
      <coord axis="wdth" value="100.0"/>
      <coord axis="wght" value="300.0"/>
    </NamedInstance>
```

but a bad fvar table with gftools-builder. So this tells me the problem is with gftools-fix-font. `gftools-fix-font` is calling `build_fvar_instances` in axis registry, so the problem is there.

```
    fvar_dflts = _fvar_dflts(ttFont)
    if not axis_dflts:
        axis_dflts = {k: v["value"] for k, v in fvar_dflts.items()}
```

This suggests to me that, unless you tell it otherwise, `axisregistry` believes that the default coordinate for the axis in an *instance* is the same as the default coordinate for that axis in the fvar axes definition. But in this case that's not true: The axis default is 112.5 but we want the instance coordinates to be 100.

*Hmm*. I wonder if this is something we support? I think for GF we want the default position of the wght axis to be 100. The [guide](https://googlefonts.github.io/gf-guide/variable.html#font-zero-origin) talks about this and gives you two tricks to work around the problem but then says it doesn't recommend them, but it doesn't say what it does recommend. It hints - but does not say explicitly - that the font origin should be at an instance location. So maybe the font is wrong here?

But let's say we do want a 112.5 axis default but a 100 width instance. Can we do it? It looks like we have to supply an `axis_dflts` parameter to axis registry, which it gets from `gftools.fix.fix_fvar_instances`, which itself is passed in as the parameter `fvar_instance_axis_dflts` in `gftools.fix.fix_font`, and *that* is passed as the `--fvar-instance-axis-dflts` parameter to `gftools-fix-font`. And *that* gets passed to `gftools-fix-font` from the builder if we set `fvarInstanceAxisDflts` in the `config.yaml`.

So if we really want to do this, we should add:

```yaml
fvarInstanceAxisDflts: wdth=100
```

to the `config.yaml`. Let's try that...

```
ValueError: Invalid configuration file
```

What? Maybe it's missing in the schema:

```python
        Optional("fvarInstanceAxisDflts"): MapPattern(Str(), Float()),
```

OK, so it's supposed to be a dictionary:

```yaml
fvarInstanceAxisDflts:
  wdth: 100
```

There, that should...

```
  File "/Users/simon/others-repos/gftools/Lib/gftools/builder/recipeproviders/googlefonts.py", line 220, in fix_args
    " --fvar-instance-axis-dflts '"
TypeError: can only concatenate str (not "dict") to str
```

You're joking, aren't you? OK, clearly nobody's ever used this facility.

Let's change it to:

```python
            args += (
                " --fvar-instance-axis-dflts '"
                + [f"{k}={v}" for k, v in self.config["fvarInstanceAxisDflts"].items()].join(" ")
                + "'"
```

Now we get

```
AttributeError: 'list' object has no attribute 'join'
```

Oh guess who has spent too much time writing Rust and not Python, where it should be

```python
                + " ".join([f"{k}={v}" for k, v in self.config["fvarInstanceAxisDflts"].items()])
```

Better!

```
  File "/opt/homebrew/lib/python3.11/site-packages/fontTools/ttLib/tables/_f_v_a_r.py", line 181, in compile
    fixedCoord = fl2fi(self.coordinates[axis], 16)
                       ~~~~~~~~~~~~~~~~^^^^^^
KeyError: 'opsz'
```

Not better? Do we have to specify all the axis defaults?

```yaml
fvarInstanceAxisDflts:
  wdth: 100
  opsz: 5
  wght: 400 # or maybe 300?
```

Success!

```xml
    <!-- Light -->
    <NamedInstance flags="0x0" subfamilyNameID="263">
      <coord axis="opsz" value="5.0"/>
      <coord axis="wdth" value="100.0"/>
      <coord axis="wght" value="300.0"/>
    </NamedInstance>
```

---

So in summary:

* gftools-fix-font is doing the right thing with an unusual font.
* I am not really sure that we want to support having an `fvar` axis default value for `wdth` (112) which is different to the coordinates of the default (Regular) instance (100). I believe GF variable fonts should have their default values "sit" on an instance. Even if we make this work, fontbakery will probably be mad at you.
* If we do want to do it, we need to add this to `config.yaml`:

```yaml
fvarInstanceAxisDflts:
  wdth: 100
  opsz: 5
  wght: 400 # or maybe 300?
```

* Nobody has done this before and it's currently broken in the builder but I have a fix which I'll PR now.

**Comment by emmamarichal** [comment]
@simoncozens 

Thank you very much for all the details and explanation! That helps a lot 🙏

We don't have any Regular masters, and we accepted to onboard the fonts like this, with all the fails related (check the last fontbakery report for the first release: https://github.com/google/fonts/pull/6374#issuecomment-1591499914)

That's why we have a font origin at opsz=5 wdth=112,5 (it's one of the masters we chose), but we wanted a result more 'in between' for the fvar (opsz=12, wdth=100).
 But I didn't have that issue last time, and that why I was surprised, since I didn't manage to reproduce the build process.

I'll try with the new version you just made, thanks a lot, I'll let you know :)

**Comment by simoncozens** [comment]
No problem, glad it is working for you! The reason it worked before is that in the past, the gftools-fix-font script did rewrite the fvar table; but we found that many fonts had extra instances or things at locations we didn't support, so we decided the best way to handle it is just to throw out the user's instances and add our own. But that only works well for "normal" fonts. :-)
